### PR TITLE
Add options to enable sanitizers

### PR DESCRIPTION
**Pull request by sakertooth**
Add options to enable AddressSanitizer, ThreadSanitizer, MemorySanitizer, and UndefinedBehaviorSanitizer. Useful for debugging memory issues, data races, etc. More information on them can be found [here](https://github.com/google/sanitizers/wiki).

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -78,6 +78,9 @@ OPTION(WANT_VST_32	"Include 32-bit VST support" ON)
 OPTION(WANT_VST_64	"Include 64-bit VST support" ON)
 OPTION(WANT_WINMM	"Include WinMM MIDI support" OFF)
 OPTION(WANT_DEBUG_FPE	"Debug floating point exceptions" OFF)
+option(WANT_ASAN	"Enable AddressSanitizer" OFF)
+option(WANT_TSAN	"Enable ThreadSanitizer" OFF)
+option(WANT_MSAN	"Enable MemorySanitizer" OFF)
```
Comment (Veratil) [inline CMakeLists.txt:83]: Let's specify that these are technically debugging options?
```suggestion
option(WANT_DEBUG_ASAN	"Enable AddressSanitizer" OFF)
option(WANT_DEBUG_TSAN	"Enable ThreadSanitizer" OFF)
option(WANT_DEBUG_MSAN	"Enable MemorySanitizer" OFF)
```

**Comment by DomClark** [comment]
It's not particularly idiomatic modern CMake to fiddle with the `CMAKE_<LANG>_FLAGS` variables. `add_compile_options` and `add_link_options` are generally preferred instead. Also, is there a particular reason to restrict the changes to the `src` directory, and not include plugins and tests too?

**Comment by sakertooth** [comment]
> It's not particularly idiomatic modern CMake to fiddle with the `CMAKE_<LANG>_FLAGS` variables. `add_compile_options` and `add_link_options` are generally preferred instead.

For some reason, these flags failed at the linking stage without setting ``CMAKE_C_FLAGS`` and ``CMAKE_CXX_FLAGS``. Before, I was trying to use ``target_link_libraries`` by passing in the flag and the other ``target_*`` functions, didn't work. Setting properities on ``lmmsobjs`` also didn't work. However, I haven't tried ``add_complie_options`` and ``add_link_options`` yet, so I'll try that since it's more idiomatic (functionally speaking though, do they act the same or different? it seems like they act the same).

> Also, is there a particular reason to restrict the changes to the `src` directory, and not include plugins and tests too?

Not really. I was under the assumption that I only had to apply the changes to the ``lmmsobjs`` target? I'll move the changes to the ``CMakeLists.txt`` at the root of the repository then. That would suffice right?

**Comment by sakertooth** [comment]
After making the changes to use ``add_compile_options`` and ``add_link_options``, and then testing to see if it builds with ``-fsanitize=thread``, the build fails. I think we need to set the ``CMAKE_<LANG>_FLAGS`` for this, as I am getting undefined references for the veal submodule inside ``LadspaEffect`` otherwise. Maybe it has something to do with ``CMAKE_<LANG>_FLAGS`` being used in all invocations of the compiler, and not just for our targets? I think this is why because with ``-DLMMS_MINIMAL=ON``, everything works fine.

**Comment by PhysSong** [comment]
I think Undefined Behavior Sanitizer(UBSAN) is also worth adding.
MSVC note: at this moment, [MSVC only supports ASAN](https://learn.microsoft.com/en-us/cpp/sanitizers/asan?view=msvc-170), but it may also support other sanitizers in the future.

**Comment by PhysSong** [comment]
> I think we need to set the `CMAKE_<LANG>_FLAGS` for this, as I am getting undefined references for the veal submodule inside `LadspaEffect` otherwise.

Could you provide your compiler version and the error message?

**Comment by sakertooth** [comment]
> > I think we need to set the `CMAKE_<LANG>_FLAGS` for this, as I am getting undefined references for the veal submodule inside `LadspaEffect` otherwise.
> 
> Could you provide your compiler version and the error message?
```
[saker@sakers-laptop ~]$ clang++ -v
clang version 16.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-pc-linux-gnu/13.2.1
Found candidate GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/13.2.1
Selected GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/13.2.1
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Selected multilib: .;@m64
```

```
cmt.cpp:(.text.tsan.module_ctor+0x5): undefined reference to `__tsan_init'
/usr/bin/ld: plugins/LadspaEffect/cmt/CMakeFiles/cmt.dir/cmt/src/delay.cpp.o: in function `initialise_delay()':
/home/saker/Git/lmms/plugins/LadspaEffect/cmt/cmt/src/delay.cpp:(.text+0x1d): undefined reference to `__tsan_func_entry'
/usr/bin/ld: /home/saker/Git/lmms/plugins/LadspaEffect/cmt/cmt/src/delay.cpp:258:(.text+0x32): undefined reference to `__tsan_memcpy'
/usr/bin/ld: /home/saker/Git/lmms/plugins/LadspaEffect/cmt/cmt/src/delay.cpp:262:(.text+0x47): undefined reference to `__tsan_memcpy'
/usr/bin/ld: /home/saker/Git/lmms/plugins/LadspaEffect/cmt/cmt/src/delay.cpp:266:(.text+0x5c): undefined reference to `__tsan_memcpy'
/usr/bin/ld: /home/saker/Git/lmms/plugins/LadspaEffect/cmt/cmt/src/delay.cpp:271:(.text+0x71): undefined reference to `__tsan_memcpy'
/usr/bin/ld: /home/saker/Git/lmms/plugins/LadspaEffect/cmt/cmt/src/delay.cpp:278:(.text+0x89): undefined reference to `__tsan_memcpy'
/usr/bin/ld: /home/saker/Git/lmms/plugins/LadspaEffect/cmt/cmt/src/delay.cpp:359:(.text+0x363): undefined reference to `__tsan_func_exit'
/usr/bin/ld: /home/saker/Git/lmms/plugins/LadspaEffect/cmt/cmt/src/delay.cpp:308:(.text+0x395): undefined reference to `__tsan_func_exit'
/usr/bin/ld: /home/saker/Git/lmms/plugins/LadspaEffect/cmt/cmt/src/delay.cpp:(.text+0x3b4): undefined reference to `__tsan_func_exit'
/usr/bin/ld: plugins/LadspaEffect/cmt/CMakeFiles/cmt.dir/cmt/src/delay.cpp.o: in function `runSimpleDelayLine(void*, unsigned long)':
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -565,6 +568,21 @@ ELSE()
 	SET (STATUS_DEBUG_FPE "Disabled")
 ENDIF(WANT_DEBUG_FPE)
 
+if(WANT_DEBUG_ASAN)
+	add_compile_options(-fsanitize=address)
+	add_link_options(-fsanitize=address)
```
Comment (Rossmaxx) [inline CMakeLists.txt:573]: ASAN on MSVC works when added to the compiler but it doesn't support linking. Will lead to an error. You don't need the linker flags on msvc.
```suggestion
    if(NOT MSVC)
	    add_link_options(-fsanitize=address)
	endif()
```

Comment (sakertooth) [reply CMakeLists.txt:573]: Better to support MSVC instead.

**Comment by messmerd** [comment]
Could we have status messages when sanitizers are enabled, similar to STATUS_DEBUG_FPE?

Something like this to set them:
```cmake
if(WANT_DEBUG_TSAN)
	if(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
		set(STATUS_DEBUG_TSAN "Enabled")
	else()
		set(STATUS_DEBUG_TSAN "Wanted but disabled due to unsupported compiler")
	endif()
else()
	set(STATUS_DEBUG_TSAN "Disabled")
endif()
```
Then the message would be displayed like this:
```cmake
MESSAGE(
"Developer options\n"
"---------------------------------------------------\n"
"* Debug FP exceptions         : ${STATUS_DEBUG_FPE}\n"
"* Debug AddressSanitizer      : ${STATUS_DEBUG_ASAN}\n"
"* Debug ThreadSanitizer       : ${STATUS_DEBUG_TSAN}\n"
"* Debug MemorySanitizer       : ${STATUS_DEBUG_MSAN}\n"
"* Debug UB Sanitizer          : ${STATUS_DEBUG_UBSAN}\n"
)
```

File: CMakeLists.txt (hunk 2 above)
Comment (sakertooth) [reply CMakeLists.txt]: We ran into some other issues with MSVC on Discord, so I'm not sure if I should still apply this suggesstion?

Comment (Rossmaxx) [reply CMakeLists.txt]: Feel free to discard. My issues were unrelated to the linker flag.

**Comment by messmerd** [comment]
Here are the results when I configured and built lmms with different combinations of settings.

```
Works?
 ______________________
|       | gcc  | clang |
|-------+------+-------|
| ASAN  | yes  |  yes  |
| TSAN  | yes  |  yes  |
| MSAN  |  1   |  yes  |
| UBSAN |  2   |   3   |
 -------+------+-------
```
As you can see, ASAN and TSAN builds work for both gcc and clang, MSAN works for clang, and the rest fail with error messages that I've pasted below.

I'm using:
**Linux Mint 21.1
gcc 11.3.0
clang 14.0.0**

I run CMake like this:
```
cmake .. -DCMAKE_INSTALL_PREFIX=../target/ -DCMAKE_PREFIX_PATH=/home/dalton/Qt/5.15.2/gcc_64/lib/cmake -DWINE_INCLUDE_DIRS=/usr/include/wine/wine -DCMAKE_PREFIX_PATH=/usr/lib/wine -DCMAKE_BUILD_TYPE=Debug -DWANT_DEBUG_ASAN=ON -DWANT_DEBUG_TSAN=OFF -DWANT_DEBUG_MSAN=OFF -DWANT_DEBUG_UBSAN=OFF -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++
```


One thing to note is that it doesn't rebuild the plugins when switching to different sanitizers but does rebuild the core.
So the sanitizers might not be set up properly for plugins. Each plugin has its own CMake target, so maybe BuildPlugin.cmake will need to be modified to support sanitizers.

**Error messages:**

**1)**
g++: error: unrecognized argument to ‘-fsanitize=’ option: ‘memory’


**2)**
[ 99%] Linking CXX executable ../lmms
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data.rel+0xd8): undefined reference to `typeinfo for lmms::PeakControllerEffect'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data.rel+0x118): undefined reference to `typeinfo for lmms::PeakControllerEffect'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data.rel+0x2d8): undefined reference to `typeinfo for lmms::PeakControllerEffect'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data.rel+0x358): undefined reference to `typeinfo for lmms::PeakControllerEffect'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data.rel+0x6d8): undefined reference to `typeinfo for lmms::PeakControllerEffect'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data.rel+0x7d8): more undefined references to `typeinfo for lmms::PeakControllerEffect' follow
collect2: error: ld returned 1 exit status
make[2]: *** [src/CMakeFiles/lmms.dir/build.make:623: lmms] Error 1
make[1]: *** [CMakeFiles/Makefile2:2161: src/CMakeFiles/lmms.dir/all] Error 2
make: *** [Makefile:156: all] Error 2

**3)**
[ 99%] Linking CXX executable ../lmms
/usr/bin/ld: /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x23
CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data+0x1d8): undefined reference to `typeinfo for lmms::PeakControllerEffect'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data+0x298): undefined reference to `typeinfo for lmms::PeakControllerEffect'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data+0x698): undefined reference to `typeinfo for lmms::PeakControllerEffect'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data+0x7b8): undefined reference to `typeinfo for lmms::PeakControllerEffect'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data+0x938): undefined reference to `typeinfo for lmms::PeakControllerEffect'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data+0x1068): more undefined references to `typeinfo for lmms::PeakControllerEffect' follow
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data+0x1b98): undefined reference to `typeinfo for lmms::PeakControllerEffectControls'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data+0x1be8): undefined reference to `typeinfo for lmms::PeakControllerEffect'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data+0x1c18): undefined reference to `typeinfo for lmms::PeakControllerEffectControls'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data+0x1cb8): undefined reference to `typeinfo for lmms::PeakControllerEffect'
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data+0x1e88): undefined reference to `typeinfo for lmms::PeakControllerEffect'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/CMakeFiles/lmms.dir/build.make:623: lmms] Error 1
make[1]: *** [CMakeFiles/Makefile2:2161: src/CMakeFiles/lmms.dir/all] Error 2
make: *** [Makefile:156: all] Error 2

**Comment by sakertooth** [comment]
> g++: error: unrecognized argument to ‘-fsanitize=’ option: ‘memory’

Did some searching and figured out GCC does not support MSAN. Will unlist it as a supported compiler from MemorySanitizer in the next few commits.

> [99%] Linking CXX executable ../lmms
/usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data.rel+0xd8): undefined reference to typeinfo for lmms::PeakControllerEffect' /usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data.rel+0x118): undefined reference to typeinfo for lmms::PeakControllerEffect'

So enabling ``CMAKE_EXPORT_COMPILE_COMMANDS`` shows that the plugins were infact not compiled with the ``-fsanitize=<check>``.  This also includes the test files, and only the files inside the ``lmmsobjs`` target were considered (core/gui files).

Using the ``add_<compile|link>_options`` commands before the CMake targets are added does add the ``-fsanitize=<check>`` as a compiler flag for all the necessary files, although I ran into linker errors with this approach, so I assumed we had to focus on compiling/linking the flag for specific targets instead. This seems to be wrong, especially since the sanitizers [expect the entirety of the program to be instrumented](https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual#non-instrumented-code). With that being said though, I'm still getting undefined reference errors for the ``cmt`` library inside the ``plugins/LadspaEffect`` directory when using ``add_<compile|link>_options``, which are actually the CMake commands we want to use without overcomplicating things.

For some reason, that ``cmt`` (and ``veal``) library is being treated separately and is not linked with the sanitizer's runtime library. If I can figure that out, I think I can get it to work.

As I was typing this, I figured it out (somewhat). Looking inside the ``cmt`` and ``veal`` ``CMakeLists.txt`` file, I found this line (using the ``veal`` target for the example, but ``cmt`` also contains something similar):

``SET_TARGET_PROPERTIES(veal PROPERTIES LINK_FLAGS "${LINK_FLAGS} -shared -Wl,-no-undefined")``

After commenting this line out, ``cmake --build . --target all`` completes with no linker errors. I think because we set the sanitizer linker flag *before* these targets were ever considered by CMake, this command overwrote and removed our ``-fsanitize=<check>`` flag, so these libraries couldn't find the sanitizer runtime library to link with. Maybe this should be fixed upstream?

**Comment by sakertooth** [comment]
> /usr/bin/ld: /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x23
CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data+0x1d8): undefined reference to typeinfo for lmms::PeakControllerEffect' /usr/bin/ld: CMakeFiles/lmmsobjs.dir/core/PeakController.cpp.o:(.data+0x298): undefined reference to typeinfo for lmms::PeakControllerEffect'

So, switching on UBSan with ``-fsanitize=undefined`` actually turns on a bunch of checks packed into UBSan, one of them being ``vptr``, which checks if objects are not of the correct dynamic type.

[From the UBSan documentation provided by Clang](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#introduction):
``
-fsanitize=vptr: Use of an object whose vptr indicates that it is of the wrong dynamic type, or that its lifetime has not begun or has ended. Incompatible with -fno-rtti. Link must be performed by clang++, not clang, to make sure C++-specific parts of the runtime library and C++ standard libraries are present.
``

As stated, ``vptr`` requires RTTI, and turning it off with ``-fno-sanitize=vptr`` makes the link sucessful. My guess is that UBSan does some instrumentation that makes ``PeakControllerEffect`` require RTTI, even when before this didn't seem to be the case.

**Review by messmerd — APPROVED** [review]
LGTM

File: CMakeLists.txt [hunk 3]
```diff
@@ -811,11 +811,11 @@ MESSAGE(
 MESSAGE(
 "Developer options\n"
 "-----------------------------------------\n"
-"* Debug FP exceptions         : ${STATUS_DEBUG_FPE}\n"
-"* Debug AddressSanitizer      : ${STATUS_DEBUG_ASAN}\n"
-"* Debug ThreadSanitizer       : ${STATUS_DEBUG_TSAN}\n"
-"* Debug MemorySanitizer       : ${STATUS_DEBUG_MSAN}\n"
-"* Debug UBSanitizer           : ${STATUS_DEBUG_UBSAN}\n"
+"* Debug using FP exceptions         : ${STATUS_DEBUG_FPE}\n"
```
Comment (DomClark) [inline CMakeLists.txt:814]: I think this one's fine to leave as is, but it doesn't matter too much - I think the floating-point exceptions can be classified as both a tool (in that they don't normally occur, and are specifically turned on to find potentially problematic calculations), and the thing being debugged (in that you look at the stack trace of them as you would any other exception).

**Review by DomClark — COMMENTED** [review]
This generally looks good to me now. One final question: what's the reason for replacing `-no-undefined` with `$<IF:$<PLATFORM_ID:GNU>,-no-undefined,-undefined>`?

**Comment by sakertooth** [comment]
> This generally looks good to me now. One final question: what's the reason for replacing `-no-undefined` with `$<IF:$<PLATFORM_ID:GNU>,-no-undefined,-undefined>`?

I made these changes when I was trying to make the project compile with ``calf`` and ``veal``, but trying it now and remove the platform check, it seems to compile fine, so I'll try to remove it.

File: CMakeLists.txt (hunk 3 above)
Comment (sakertooth) [reply CMakeLists.txt:814]: I changed it for consistency. I didn't like how it wasn't alligned with the other status messages. If the preference is towards leaving it be, I don't have a problem with that.

**Review by DomClark — APPROVED** [review]

**Comment by sakertooth** [comment]
[a55502e](https://github.com/LMMS/lmms/pull/6840/commits/a55502e75ae7168ed07110ec7fc25f267532faa2) broke the builds somehow (I was sure I did a clean build..). Removing the ``-no-undefined`` option works. Not too sure exactly what that flag does, but I think it was conflicting with linking the sanitizers.

I'll verify again to make sure it is working at least on GCC and Clang before merge. MSVC support seems a bit wonky right now, so I'm not sure what the best approach would be to make it work. We might just have to wait.

**Comment by DomClark** [comment]
> Not too sure exactly what that flag does, but I think it was conflicting with linking the sanitizers.

My understanding from [the man page](https://linux.die.net/man/1/ld) is that it causes references to undefined symbols (e.g. calls to undefined functions) to result in link errors, rather than trying to resolve them at run time. This sounds more like a safety check, so if the code linked correctly with the flag, then it will link and run correctly without it. I think such errors should still be caught by Windows builds anyway. (As far as I understand, Windows doesn't support linking with unresolved symbols - if you want to do this sort of thing, you need to bypass the linker and manually call `GetProcAddress`, the Windows equivalent of `dlsym`.)

**Pull request merged**